HysteresisActivity::impulse() thrashes a timer

smfr opened the pull request:
#### c408a82764d20985bb78dbc1b4847bfa9eb5dad9
<pre>
HysteresisActivity::impulse() thrashes a timer
<a href="https://bugs.webkit.org/show_bug.cgi?id=256572">https://bugs.webkit.org/show_bug.cgi?id=256572</a>
rdar://109134335

Reviewed by Chris Dumez.

`HysteresisActivity::impulse()` stops and re-starts a timer simply to reschedule it into the future;
this destroys and recreates the timer every time, which is expensive. The implementation of
`HysteresisActivity::impulse()` can simply restart the timer.

We can make this cheaper by rescheduling the CFTimerRef if it hasn&apos;t fired yet, so add support for
this in `RunLoop::TimerBase::start()`; we can only reschedule non-repeating timers which haven&apos;t
fired yet, but this is fairly common.

* Source/WTF/wtf/cf/RunLoopCF.cpp:
(WTF::RunLoop::TimerBase::start):
* Source/WebCore/PAL/pal/HysteresisActivity.h:
(PAL::HysteresisActivity::start):
(PAL::HysteresisActivity::stop):
(PAL::HysteresisActivity::impulse):

Canonical link: <a href="https://commits.webkit.org/263946@main">https://commits.webkit.org/263946@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94ec88e28f071a77387c0e54e6e567136660dff6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6173 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6361 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6543 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7736 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6508 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6173 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6581 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6313 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9387 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6283 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6300 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5587 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7803 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3774 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5567 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13461 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5186 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5639 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5646 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7882 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5768 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6118 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5003 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6294 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5535 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1498 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1469 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9679 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6470 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5902 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1625 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/PAL/pal/HysteresisActivity.h:
```diff
@@ -61,17 +61,15 @@ class HysteresisActivity {
-            stop();
-        }
+        if (!m_active)
+            m_timer.startOneShot(m_hysteresisSeconds);
```
If `!m_timer.isActive()`, it seems we fail to call `m_callback(HysteresisState::Started)` now no?

cdumez reviewed (approved):
LGTM

smfr committed:
HysteresisActivity::impulse() thrashes a timer
https://bugs.webkit.org/show_bug.cgi?id=256572
rdar://109134335

Reviewed by Chris Dumez.

`HysteresisActivity::impulse()` stops and re-starts a timer simply to reschedule it into the future;
this destroys and recreates the timer every time, which is expensive. The implementation of
`HysteresisActivity::impulse()` can simply restart the timer.

We can make this cheaper by rescheduling the CFTimerRef if it hasn't fired yet, so add support for
this in `RunLoop::TimerBase::start()`; we can only reschedule non-repeating timers which haven't
fired yet, but this is fairly common.

* Source/WTF/wtf/cf/RunLoopCF.cpp:
(WTF::RunLoop::TimerBase::start):
* Source/WebCore/PAL/pal/HysteresisActivity.h:
(PAL::HysteresisActivity::start):
(PAL::HysteresisActivity::stop):
(PAL::HysteresisActivity::impulse):

Canonical link: https://commits.webkit.org/263946@main
